Update Delete to handle long-running deletions

ekinanp opened the pull request.

ekinanp committed:
Update Delete's signature to (deleted, error)

If deleted is true, then the entry was deleted. Otherwise, the entry's
deletion is in-progress. The latter's useful when deleting e.g. VMs
because it gives us a way to let the user know that the VM's being deleted
without having to wait for the deletion, which can take a long time
(several minutes in GCP, and up to an hour in AWS).

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update the external plugin docs with delete's calling convention

We'll add delete to the action docs once delete is implemented for some
of the core plugin entries since those are what the examples are based
off of.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on docs/docs/external-plugins.md:
```diff
@@ -208,6 +209,19 @@ bash-3.2$ /path/to/myplugin.rb schema /myplugin/foo ''
+
+When `delete` is invoked, the script must output a boolean JSON. `true` means that the entry was deleted. `false` means that the entry is marked for deletion and will eventually be deleted.
+
+**Note:** If you anticipate `delete` taking a long time (> 30 seconds), then output `false`.
```
I wonder if we can use a similar pattern for `signal`.

ekinanp commented on plugin/methodWrappers.go:
```diff
@@ -139,18 +139,26 @@ func Stream(ctx context.Context, s Streamable) (io.ReadCloser, error) {
+		return
 	}
+	if !deleted {
+		// The entry will eventually be deleted so leave the cache alone.
```
Should we refresh the entry here? Thinking a `parent.List` call, though the latter could add up when deleting multiple entries from the same parent.

I lean towards leaving it alone until it becomes an issue.

MikaelSmith replied:
Instead of explicitly refreshing, I'd clear the cache entry. That delays the List call until it's actually needed, and raises the likelihood that the entry has been removed.

ekinanp replied:
Ah yeah, that is better. Thanks, will update.

MikaelSmith commented on plugin/types.go:
```diff
@@ -195,11 +195,8 @@ type Readable interface {
-// should ensure that it and all its children are removed. If the entry has
-// any dependencies that need to be deleted, then Delete should return an
-// error.
+// Deletable is an entry that can be deleted.
```
Why'd you remove the description? We also need to describe what the boolean return value represents.

ekinanp replied:
I was trying to place the plugin-author specific action info in the external plugin docs and leave the comments here to core-plugin specific subtleties. I guess it wouldn't hurt to include a brief description.

ekinanp committed:
Fix cache updates in plugin.Delete

The changes here ensure that Wash will always get fresh data if the user
has further interactions with the deleted entry or its parent.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Update delete docs

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented on plugin/types.go:
```diff
@@ -195,7 +195,8 @@ type Readable interface {
 
-// Deletable is an entry that can be deleted.
+// Deletable is an entry that can be deleted. Its semantics are described
+// in https://puppetlabs.github.io/wash/docs/external-plugins#delete
```
I think people should be able to figure out what the return values mean from the external plugin docs.

MikaelSmith replied:
Someone - including our future selves - writing core plugins might not think to go look there.

MikaelSmith replied:
The core plugin documentation should largely stand on its own.

MikaelSmith replied:
Ah, I guess you link to it. But it feels weird to me for the core plugin documentation to link out to an external page just to explain what the function does and returns.

ekinanp commented:
Updated to handle cache refreshing and also made some docs updates.

MikaelSmith commented on plugin/cache.go:
```diff
@@ -74,6 +74,22 @@ func opKeysRegex(path string) *regexp.Regexp {
+// opKeyRegex returns a regex that matches the key <op>::<path>.
+// It is useful for clearing a specific cached op for the entry
+// corresponding to <path>.
+func opKeyRegex(op string, path string) *regexp.Regexp {
```
What changed here?

ekinanp replied:
I want to clear out `List::<parent>` only in `plugin.Delete`. `opKeysRegex` clears out `Op::<parent>` and `Op::<children>`.

ekinanp committed:
Change opKeysRegex => allOpKeysIncludingChildrenRegex

Latter better reflects what the regex matches. This change also moves
allOpKeysIncludingChildrenRegex closer to ClearCacheFor so the latter's
dependency on the former is more apparent.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Ok, updated again to address the feedback.
